test-collateral/generate-policy/src/main.rs

```diff
@@ -0,0 +1,638 @@
+    }
+}
+
+/// Serializes the identities of all principals in the Veracruz computation into
+/// a JSON value.
+fn serialize_identities(arguments: &Arguments) -> Value {
+    assert_eq!(&arguments.certificates.len(), &arguments.roles.len());
+
+    let mut values = vec![];
+
+    for (id, (cert, roles)) in arguments
+        .certificates
+        .iter()
+        .zip(&arguments.roles)
+        .enumerate()
+    {
+        if let Ok(mut file) = File::open(cert) {
+            let mut content = String::new();
+
+            file.read_to_string(&mut content)
+                .expect("Failed to read file.");
+
+            content = content.replace(
+                "-----BEGIN CERTIFICATE-----",
+                "-----BEGIN CERTIFICATE-----\\n",
+            );
+            content = content.replace("-----END CERTIFICATE", "\\n-----END CERTIFICATE");
+
+            let json = json!({
+                "certificate": content,
+                "id": id,
+                "roles": roles,
+            });
+
+            values.push(json);
+        } else {
+            abort_with("Could not open certificate file.");
+        }
+    }
+
+    json!(values)
+}
+
+/// Serializes the enclave server certificate expiry timepoint to a JSON value,
+/// computing the time when the certificate will expire as a point relative to
+/// the current time.
+fn serialize_enclave_certificate_expiry(arguments: &Arguments) -> Value {
+    let expiry = arguments
+        .certificate_expiry
+        .expect("Internal invariant failed: certificate lifetime is missing.");
+
+    json!({
+        "year": expiry.year(),
+        "month": expiry.month(),
+        "day": expiry.day(),
+        "hour": expiry.hour(),
+        "minute": expiry.minute()
+    })
+}
+
+/// Serializes the Veracruz policy file as a JSON value.
+///
+/// NOTE: we are glossing over TrustZone attestation for the moment, so we use
+/// the measurement of the SGX enclave as the measurement of the TrustZone
+/// trusted application, too.
+fn serialize_json(arguments: &Arguments) -> Value {
+    info!("Serializing JSON policy file.");
+
+    json!({
+        "identities": serialize_identities(arguments),
+        "sinaloa_url": format!("{}", &arguments.sinaloa_url.as_ref().unwrap()),
+        "enclave_cert_expiry": serialize_enclave_certificate_expiry(arguments),
+        "ciphersuite": POLICY_CIPHERSUITE,
+        "tabasco_url": format!("{}", &arguments.tabasco_url.as_ref().unwrap()),
+        "data_provision_order": json!(&arguments.data_provisioning_order),
+        "streaming_order": json!(&arguments.streaming_provisioning_order),
+        "pi_hash": compute_program_hash(arguments),
+        "debug": &arguments.enclave_debug_mode,
+        "execution_strategy": &arguments.execution_strategy,
+        "mexico_city_hash_nitro": compute_nitro_enclave_hash(arguments),
+        "mexico_city_hash_sgx": compute_sgx_enclave_hash(arguments),
+        "mexico_city_hash_tz": compute_sgx_enclave_hash(arguments),
```

Comment:
Each of these field elements, mexico_city_hash_nitro, mexico_city_hash_sgx, and mexico_city_hash_tz, are all options. Or more accurately, one or more of them must be present, but they do not all need to be present.

Reply:
Will fix that.